File: src/domain/metricsEngine.ts
// Habitat Metrics, Resistance & Gamification Engine
import { DisciplineMode } from './types';

export interface XpCalculationInput {
  baseXp: number;
  resistanceSeconds: number;
  attemptsCount: number;
  disciplineMode: DisciplineMode;
}

export interface XpCalculationResult {
  totalXp: number;
  speedMultiplier: number;
  modeMultiplier: number;
  firstAlarmBonus: boolean;
  resistanceMinutes: number;
}

export class MetricsEngine {
  /**
   * Calculates resistance in seconds between scheduled/triggered time and completion
   */
  public static calculateResistanceSeconds(startTime: Date | string, completionTime: Date | string): number {
    const start = new Date(startTime).getTime();
    const end = new Date(completionTime).getTime();
    return Math.max(0, Math.round((end - start) / 1000));
  }

  /**
   * Computes XP earned for a completed mission with speed and strictness multipliers
   */
  public static calculateXp(input: XpCalculationInput): XpCalculationResult {
    const { baseXp, resistanceSeconds, attemptsCount, disciplineMode } = input;
    const resistanceMinutes = resistanceSeconds / 60;

    // Speed multiplier logic
    let speedMultiplier = 1.0;
    let firstAlarmBonus = false;

    if (attemptsCount === 1 && resistanceMinutes <= 2) {
      speedMultiplier = 1.5; // +50% Instant Action Bonus
      firstAlarmBonus = true;
    } else if (attemptsCount === 1 && resistanceMinutes <= 5) {
      speedMultiplier = 1.0; // Nominal speed
      firstAlarmBonus = true;
    } else {
      // Penalize procrastination retries
      speedMultiplier = Math.max(0.5, 1.0 - (attemptsCount - 1) * 0.15);
    }

    // Strictness mode multiplier
    let modeMultiplier = 1.0;
    switch (disciplineMode) {
      case 'HARDCORE':
        modeMultiplier = 1.3;
        break;
      case 'DISCIPLINE':
        modeMultiplier = 1.0;
        break;
      case 'GENTLE':
        modeMultiplier = 0.9;
        break;
    }

    const totalXp = Math.round(baseXp * speedMultiplier * modeMultiplier);

    return {
      totalXp,
      speedMultiplier,
      modeMultiplier,
      firstAlarmBonus,
      resistanceMinutes: parseFloat(resistanceMinutes.toFixed(2))
    };
  }

  /**
   * Calculates new Discipline Score (0 - 100 scale) based on recent mission performance
   */
  public static calculateUpdatedDisciplineScore(
    currentScore: number,
    completed: boolean,
    resistanceMinutes: number,
    mode: DisciplineMode
  ): number {
    let delta = 0;

    if (completed) {
      if (resistanceMinutes <= 2) delta = +4;
      else if (resistanceMinutes <= 5) delta = +2;
      else delta = +1;

      if (mode === 'HARDCORE') delta += 1;
    } else {
      // Missed mission penalty
      delta = mode === 'HARDCORE' ? -6 : -4;
    }

    return Math.max(0, Math.min(100, currentScore + delta));
  }

  /**
   * Streak and Grace Protocol Evaluator
   */
  public static evaluateStreak(
    currentStreak: number,
    graceTokens: number,
    missionSuccess: boolean
  ): {
    newStreak: number;
    newGraceTokens: number;
    usedGraceToken: boolean;
    streakBroken: boolean;
  } {
    if (missionSuccess) {
      const newStreak = currentStreak + 1;
      // Award 1 grace token every 14 days of streak (max 3 tokens)
      const earnedGrace = newStreak % 14 === 0 && graceTokens < 3;
      return {
        newStreak,
        newGraceTokens: earnedGrace ? graceTokens + 1 : graceTokens,
        usedGraceToken: false,
        streakBroken: false
      };
    } else {
      // If user failed/missed mission
      if (graceTokens > 0) {
        // Protect streak with Grace Token
        return {
          newStreak: currentStreak,
          newGraceTokens: graceTokens - 1,
          usedGraceToken: true,
          streakBroken: false
        };
      } else {
        // Streak broken
        return {
          newStreak: 0,
          newGraceTokens: 0,
          usedGraceToken: false,
          streakBroken: true
        };
      }
    }
  }

  /**
   * Autonomy Score (0 - 100): Measures internal vs external habit strength
   * Higher score = less reliance on sirens and faster natural execution
   */
  public static calculateAutonomyScore(
    avgRecentResistanceMinutes: number,
    baselineResistanceMinutes: number = 15
  ): number {
    if (baselineResistanceMinutes <= 0) return 100;
    const ratio = avgRecentResistanceMinutes / baselineResistanceMinutes;
    const autonomy = (1 - Math.min(1, Math.max(0, ratio))) * 100;
    return Math.round(autonomy);
  }
}

File: packages/discipline_engine/lib/src/metrics/metrics_engine.dart
// Metrics, XP Formulas, Resistance (ΔtR), and Streak Grace Engine
import '../models/alarm.dart';

class XpReward {
  final int totalXp;
  final double speedMultiplier;
  final double modeMultiplier;
  final bool isFirstAlarmBonus;
  final double resistanceMinutes;

  const XpReward({
    required this.totalXp,
    required this.speedMultiplier,
    required this.modeMultiplier,
    required this.isFirstAlarmBonus,
    required this.resistanceMinutes,
  });
}

class StreakEvaluationResult {
  final int newStreak;
  final int newGraceTokens;
  final bool usedGraceToken;
  final bool streakBroken;

  const StreakEvaluationResult({
    required this.newStreak,
    required this.newGraceTokens,
    required this.usedGraceToken,
    required this.streakBroken,
  });
}

class MetricsEngine {
  /// Resistance ΔtR = t_completed - t_scheduled (in seconds)
  static int calculateResistanceSeconds({
    required DateTime startTime,
    required DateTime completionTime,
  }) {
    final diff = completionTime.difference(startTime).inSeconds;
    return diff < 0 ? 0 : diff;
  }

  /// Calculates XP earned with Speed & Strictness Multipliers
  static XpReward calculateXp({
    required int baseXp,
    required int resistanceSeconds,
    required int attemptCount,
    required DisciplineMode disciplineMode,
  }) {
    final resistanceMinutes = resistanceSeconds / 60.0;
    double speedMultiplier = 1.0;
    bool isFirstAlarmBonus = false;

    if (attemptCount == 1 && resistanceMinutes <= 2.0) {
      speedMultiplier = 1.5; // +50% Instant Action Bonus
      isFirstAlarmBonus = true;
    } else if (attemptCount == 1 && resistanceMinutes <= 5.0) {
      speedMultiplier = 1.0; // Nominal speed
      isFirstAlarmBonus = true;
    } else {
      // Procrastination retry penalty
      speedMultiplier = (1.0 - (attemptCount - 1) * 0.15).clamp(0.5, 1.0);
    }

    double modeMultiplier = switch (disciplineMode) {
      DisciplineMode.hardcore => 1.3,
      DisciplineMode.discipline => 1.0,
      DisciplineMode.gentle => 0.9,
    };

    final totalXp = (baseXp * speedMultiplier * modeMultiplier).round();

    return XpReward(
      totalXp: totalXp,
      speedMultiplier: speedMultiplier,
      modeMultiplier: modeMultiplier,
      isFirstAlarmBonus: isFirstAlarmBonus,
      resistanceMinutes: double.parse(resistanceMinutes.toStringAsFixed(2)),
    );
  }

  /// Evaluates Streak and Grace Token Vault
  static StreakEvaluationResult evaluateStreak({
    required int currentStreak,
    required int graceTokens,
    required bool missionSuccess,
  }) {
    if (missionSuccess) {
      final newStreak = currentStreak + 1;
      // Award 1 grace token every 14 days of consistency (max 3)
      final earnedGrace = (newStreak % 14 == 0) && (graceTokens < 3);
      return StreakEvaluationResult(
        newStreak: newStreak,
        newGraceTokens: earnedGrace ? graceTokens + 1 : graceTokens,
        usedGraceToken: false,
        streakBroken: false,
      );
    } else {
      if (graceTokens > 0) {
        // Protect streak using Grace Token
        return StreakEvaluationResult(
          newStreak: currentStreak,
          newGraceTokens: graceTokens - 1,
          usedGraceToken: true,
          streakBroken: false,
        );
      } else {
        // Streak Broken
        return const StreakEvaluationResult(
          newStreak: 0,
          newGraceTokens: 0,
          usedGraceToken: false,
          streakBroken: true,
        );
      }
    }
  }

  /// Autonomy Score (0 - 100): External -> Internal habit transition index
  static int calculateAutonomyScore({
    required double avgRecentResistanceMinutes,
    double baselineResistanceMinutes = 15.0,
  }) {
    if (baselineResistanceMinutes <= 0) return 100;
    final ratio = (avgRecentResistanceMinutes / baselineResistanceMinutes).clamp(0.0, 1.0);
    return ((1.0 - ratio) * 100).round();
  }
}
